### PR TITLE
Add Wireless Network support with dedicated buildable

### PR DESCRIPTION
**Pull request by rockfactory**
This PR introduces Wireless Network Communications through a new buildable, the Wireless Access Point (WAP), which can be connected to a Radar Tower in order to receive and transmit messages over the air. 

- The WAP has an easy to use UI where pioneers can check their wireless network and detect issues, like missing Access Points, too far towers and so on.
- Messages could be transmitted only to towers which are connected to a WAP and are in range. The range is calculated using the same Formula from SF for scanning, so an higher tower has an higher range.
- The Wireless Access Point requires Crystal Oscillators to be built, keeping a parallelism with SF recipes where Oscillator is used

On the technical side:
- The `WirelessAccessPoint` code is based on the `NetworkRouter` with some adaptation. This means loops are prevented in the same way.
- The main optimization is that we cache the available connections from each Access Point (using the WirelessSubsystem which is introduced in this commit too). Even if the alghoritm itself is not optimized, we need to update connections only when a WAP is built / destroyed, so very unfrequently.

TODOs/bugs - maybe other PRs based even on @Panakotta00 / community feeling:
- [x] Destroying a Radar Tower will not immediately disable communications with its connected WAP. I'd like to implement auto-dismantle so while removing the tower, you remove the WAP too.
- [x] ~~Currently WAP is unlocked with the same schematic for the Network Router. We could add a "Wireless Networking" schematic in the same Tier, but I feel it would be a bit lonely~~ too lonely
- [x] Radar Tower does not require power in order to send messages
- [x] Clicking on a Tower in the WAP interact widget opens the Marker Editor
- [x] Wireless Access Point cannot be rotated around the Radar Tower (fixed positioning while snapped)
- [x] Check for multiplayer support. Replication should be working, but it requires testing. _Update: tested with Chaka, it works!_
- [x] Map panning with rmb doesn't work

An UI screen:
Satisfactory Screenshot 2022 09 10 - 17 25 19 07

### CODE REVIEW

**Comment by rockfactory** [comment]
I'll share here a link to an alpha build and a temporary documentation if someone likes to test this version and help us checking everything is fine, especially in multiplayer:

Steps:
1. Research Network Connectors, it will unlock Wireless Access Point too
2. Build a radar tower, then build a wireless access point. it will snap to the tower
3. The WAP has only one connection, connect it to your local/router network
4. Do the same with another radar tower & WAP
5. Use "E" to interact with the access point - it will show a map of other radar towers and connection status, like signal strength, presence of an access point, etc. I think it's really useful to double check your network! Bonus: you can click on a radar tower to highlight in on the map. Give the radar towers a custom name (interacting with the tower itself), and its name will show on the WAP map
6. Now you should be able to send messages between the network! Remember only network messages are allowed, refer to same docs: https://docs.ficsit.app/ficsit-networks/latest/lua/guide/InterComputerCommunication.html

**Comment by rockfactory** [comment]
Released another alpha (`alpha.2`, link above), which:
- Should fix a bug with multiplayer
- Fix map panning with right mouse button
- Radar Towers now require power to operate the attached Access Point

**Comment by rockfactory** [comment]
Aaand [`alpha.3`](https://1drv.ms/u/s!AqQ0YW9so_5OgUPPU0CCh7xgR5UN?e=8tGol0), which allows WAP rotation with the scroll wheel plus some UI enhancements!

**Comment by rockfactory** [comment]
Well, some updates. We did a lot of testing, even on Multiplayer, with Chaka (thanks for your patience!) and we fixed some bugs, plus rewrote the multiplayer bits in order to make sure it works correctly. [`alpha.6`](https://1drv.ms/u/s!AqQ0YW9so_5OgUh4j9k3G5rJUXPF?e=XLRMtO) is ready!

File: Source/FicsItNetworks/Network/Wireless/FINWirelessAccessPointConnectionData.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿// Copyright Coffee Stain Studios. All Rights Reserved.
```
Comment (Panakotta00) [inline Source/FicsItNetworks/Network/Wireless/FINWirelessAccessPointConnectionData.h:1]: Mostly not true, right?

File: Source/FicsItNetworks/Network/Wireless/FINWirelessRCO.cpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright Coffee Stain Studios. All Rights Reserved.
```
Comment (Panakotta00) [inline Source/FicsItNetworks/Network/Wireless/FINWirelessRCO.cpp:1]: Untrue

File: Source/FicsItNetworks/Network/Wireless/FINWirelessRCO.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright Coffee Stain Studios. All Rights Reserved.
```
Comment (Panakotta00) [inline Source/FicsItNetworks/Network/Wireless/FINWirelessRCO.h:1]: untrue

File: Source/FicsItNetworks/Network/Wireless/FINWirelessSubsystem.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
+﻿// Copyright Coffee Stain Studios. All Rights Reserved.
```
Comment (Panakotta00) [inline Source/FicsItNetworks/Network/Wireless/FINWirelessSubsystem.cpp:1]: untrue

File: Source/FicsItNetworks/Network/Wireless/FINWirelessSubsystem.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
+﻿// Copyright Coffee Stain Studios. All Rights Reserved.
```
Comment (Panakotta00) [inline Source/FicsItNetworks/Network/Wireless/FINWirelessSubsystem.h:1]: Untrue

**Review by Panakotta00 — CHANGES_REQUESTED** [review]
Additionally please refrain from using the auto keyword and instead fully type them out.

Otherwise, amazing work! Well done!

File: Source/FicsItNetworks/Components/FINWirelessAccessPointHologram.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+#pragma once
+
+#include "CoreMinimal.h"
+#include "FGConstructDisqualifier.h"
+#include "GameFramework/Actor.h"
+#include "Hologram/FGBuildableHologram.h"
+#include "Hologram/FGHologram.h"
+#include "FINWirelessAccessPointHologram.generated.h"
+
+#define LOCTEXT_NAMESPACE "Construct Disqualifier FIN"
```
[the remark this answers is not in the document]
Comment (rockfactory) [reply Source/FicsItNetworks/Components/FINWirelessAccessPointHologram.h:12]: I can do it without any issue, but I copied the same structure in the original one (`Source/FactoryGame/Public/FGConstructDisqualifier.h`), e.g.:

```cpp
#define LOCTEXT_NAMESPACE "Construct Disqualifiers" 
```

```cpp
UFGConstructDisqualifier() :
		mDisqfualifyingText( LOCTEXT( "UFGConstructDisqualifier", "This should never be seen!" ) ),
		mIsSoftDisqualifier( false )
	{
	}
```

Just double checking, if you prefer to switch them I'll switch them asap

**Comment by rockfactory** [comment]
@Panakotta00 changes applied, tried to remove some of the `auto`s. Just le me know if the disqualifer loc namespace is ok 👍

File: Source/FicsItNetworks/Components/FINWirelessAccessPointHologram.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Panakotta00) [reply Source/FicsItNetworks/Components/FINWirelessAccessPointHologram.h:10]: I find that as an rather odd way of grouping but on seccond thought i agree

**Review by Panakotta00 — APPROVED** [review]

**Comment by Panakotta00** [comment]
If you want to you can now also add credit in the authors list in the .uplugin, the Readme and the docs if you want to.
(Just your social handle and a link to some preffered place of yours)

**Pull request merged**